// @nxext/stencil/src/executors/stencil-runtime/e2e-testing.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.cleanupE2eTesting = exports.prepareE2eTesting = void 0;
const fileutils_1 = require("../../utils/fileutils");
const fileutils_2 = require("@nrwl/workspace/src/utilities/fileutils");
const path_1 = require("path");
const devkit_1 = require("@nrwl/devkit");
function prepareE2eTesting(pathCollection) {
    const libPackageJson = {
        name: pathCollection.projectName,
        version: '1',
        main: `${path_1.relative(pathCollection.projectRoot, pathCollection.distDir)}/dist/index.js`,
        module: `${path_1.relative(pathCollection.projectRoot, pathCollection.distDir)}/dist/index.mjs`,
        es2015: `${path_1.relative(pathCollection.projectRoot, pathCollection.distDir)}/dist/esm/index.mjs`,
        es2017: `${path_1.relative(pathCollection.projectRoot, pathCollection.distDir)}/dist/esm/index.mjs`,
        types: `${path_1.relative(pathCollection.projectRoot, pathCollection.distDir)}/dist/types/index.d.ts`,
        collection: `${path_1.relative(pathCollection.projectRoot, pathCollection.distDir)}/dist/collection/collection-manifest.json`,
        'collection:main': `${path_1.relative(pathCollection.projectRoot, pathCollection.distDir)}/dist/collection/index.js`,
        unpkg: `${path_1.relative(pathCollection.projectRoot, pathCollection.distDir)}/dist/${pathCollection.projectName}/${pathCollection.projectName}.js`,
        files: [
            `${path_1.relative(pathCollection.projectRoot, pathCollection.distDir)}/dist/`,
            `${path_1.relative(pathCollection.projectRoot, pathCollection.distDir)}/loader/`
        ]
    };
    fileutils_2.writeJsonFile(devkit_1.joinPathFragments(`${pathCollection.projectRoot}/package.e2e.json`), libPackageJson);
}
exports.prepareE2eTesting = prepareE2eTesting;
function cleanupE2eTesting(pathCollection) {
    const pkgJsonPath = devkit_1.joinPathFragments(`${pathCollection.projectRoot}/package.e2e.json`);
    if (fileutils_2.fileExists(pkgJsonPath)) {
        fileutils_1.deleteFile(pkgJsonPath);
    }
}
exports.cleanupE2eTesting = cleanupE2eTesting;
