// @nxext/stencil/src/stencil-core-utils/lib/style-plugins.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addStylePluginToConfig = exports.addStylePlugin = exports.SupportedStyles = void 0;
const plugins_1 = require("./plugins");
const devkit_1 = require("@nrwl/devkit");
const ast_utils_1 = require("../../utils/ast-utils");
var SupportedStyles;
(function (SupportedStyles) {
    SupportedStyles["css"] = "css";
    SupportedStyles["scss"] = "scss";
    SupportedStyles["styl"] = "styl";
    SupportedStyles["less"] = "less";
    SupportedStyles["pcss"] = "pcss";
})(SupportedStyles = exports.SupportedStyles || (exports.SupportedStyles = {}));
function addStylePlugin(stencilConfigSource, style) {
    const styleImports = {
        [SupportedStyles.css]: [],
        [SupportedStyles.scss]: [
            ...ast_utils_1.addImport(stencilConfigSource, `import { sass } from '@stencil/sass'`),
            ...plugins_1.addToPlugins(stencilConfigSource, 'sass()')
        ],
        [SupportedStyles.styl]: [
            ...ast_utils_1.addImport(stencilConfigSource, `import { stylus } from '@stencil/stylus'`),
            ...plugins_1.addToPlugins(stencilConfigSource, 'stylus()')
        ],
        [SupportedStyles.less]: [
            ...ast_utils_1.addImport(stencilConfigSource, `import { less } from '@stencil/less`),
            ...plugins_1.addToPlugins(stencilConfigSource, 'less()')
        ],
        [SupportedStyles.pcss]: [
            ...ast_utils_1.addImport(stencilConfigSource, `import { postcss } from '@stencil/postcss`),
            ...ast_utils_1.addImport(stencilConfigSource, `import autoprefixer from 'autoprefixer`),
            ...plugins_1.addToPlugins(stencilConfigSource, `
          postcss({
            plugins: [autoprefixer()]
          })
          `)
        ]
    };
    return styleImports[style];
}
exports.addStylePlugin = addStylePlugin;
function addStylePluginToConfig(host, stencilConfigPath, style) {
    const stencilConfigSource = ast_utils_1.readTsSourceFile(host, stencilConfigPath);
    const changes = devkit_1.applyChangesToString(stencilConfigSource.text, addStylePlugin(stencilConfigSource, style));
    host.write(stencilConfigPath, changes);
}
exports.addStylePluginToConfig = addStylePluginToConfig;
